packages/workspace-resolver/src/resolve-workspace.ts

```diff
@@ -0,0 +1,187 @@
+    path,
+    name: json.name,
+    version: json.version,
+    workspace: !!json.workspaces,
+    children: [],
+    parent,
+    modular: {
+      type: 'unknown',
+      ...json.modular,
+    },
+    // Like yarn classic `workspaces info`, we include all except peerDependencies
+    dependencies: {
+      ...json.optionalDependencies,
+      ...json.devDependencies,
+      ...json.dependencies,
+    },
+  };
+  collector.set(json.name, pkg);
+
+  if (json.modular?.type === 'root' && !isRoot) {
+    throw new Error(
+      'Nested modular roots are currently not supported by Modular',
+    );
+  }
+
+  // Allow for the `workspaces` value to be `[]` or `{}`, otherwise throw (as nested workspaces currently unsupported)
+  if (!isRoot && json.workspaces) {
+    if (Array.isArray(json.workspaces) && json.workspaces.length > 0) {
+      throw new Error(
+        'Nested workspaces are currently not supported by Modular',
+      );
+    }
+
+    if (
+      typeof json.workspaces === 'object' &&
+      !Array.isArray(json.workspaces) &&
+      Object.keys(json.workspaces).length > 0
+    ) {
+      throw new Error(
+        'Nested workspaces are currently not supported by Modular',
+      );
+    }
+  }
+
+  for (const link of resolveWorkspacesDefinition(root, json.workspaces)) {
+    const [, child] = await resolveWorkspace(
+      workingDir,
+      false,
+      link,
+      pkg,
+      collector,
+    );
+    child && pkg.children.push(child);
+  }
+
+  return [collector, pkg];
+}
+
+export function analyzeWorkspaceDependencies(
+  workspacePackages: Map<string, ModularWorkspacePackage>,
+): WorkspaceMap {
+  const mappedDeps = new Map<string, WorkspaceObj>();
+  const exhaustivePackageNameList = Array.from(workspacePackages.keys());
+  const allPackages = Array.from(workspacePackages.entries());
+
+  // Exclude the root when analyzing package inter-dependencies
+  const packagesWithoutRoot = Array.from(workspacePackages.entries()).filter(
+    ([, pkg]) => {
+      return pkg.modular.type !== 'root';
+    },
+  );
+
+  // Calculate deps and mismatches a-la Yarn classic `workspaces info`
+  packagesWithoutRoot.forEach(([pkgName, pkg]) => {
+    const packageDepNames = Object.keys(pkg.dependencies || {}).filter(
+      (dep) => {
+        return exhaustivePackageNameList.includes(dep);
+      },
+    );
+    const packageDeps = allPackages.filter(([, pkg]) =>
+      packageDepNames.includes(pkg.name),
+    );
+
+    // Mismatched = version in packages/<package>/package.json does not satisfy the dependent's range
+    const mismatchedWorkspaceDependencies = Object.entries(
+      pkg.dependencies || {},
+    )
+      .filter(([dep, range]) => {
+        const matchingPackage = packageDeps.find(
+          ([matchingPackageName]) => dep === matchingPackageName,
+        );
+        if (!matchingPackage) {
+          return false;
+        }
+
+        const [, match] = matchingPackage;
+        return !semver.satisfies(match.version, range);
```

Comment:
Could we support [workspace ranges](https://yarnpkg.com/features/workspaces#workspace-ranges-workspace) too here? They're heavily used in Yarn >= v2 and they might be "standardised" in the future (ie supported by a npm RFC)

Reply:
Yep adding this, but: it should be noted, this makes `@modular-scripts/workspace-resolver` no longer 100% compatible with `yarn workspaces info` (which makes sense, because workspace ranges were not a thing in yarn v1).